Option to show minimized views in scale

dkondor opened the pull request:
This can be helpful if scale is used as a primary method to switch between views. Of course, this is probably highly opinion-based, so I made this into a config option. Depends on #1906 

Issues / questions:
 - [x] Test the case when a view is minimized while scale is running ~~(should work, but not easy to test, I don't find the interactive scale option anymore)~~ simple test case [here](https://github.com/dkondor/wayfire/blob/scale_test/minimize_window.py)
 - [x] Should this be a global option? Yes
 - ~~[ ] Option for an other plugin starting scale to specify whether to include minimized views (e.g. the dbus plugin [here](https://github.com/damianatorrpm/wayfire-plugin_dbus_interface/blob/c705bf1e9dbe54b7ecc3608869db7c53e82bb38d/dbus_interface_backend.cpp#L1095))~~
 - [x] Should minimized view be differentiated in some way? E.g. more faded, different title overlay, etc.
 - [ ] Any view that is focused during scale is unminimzed. This is an issue when using the arrow keys which focuses the views under selection. I see that a related issue came up in #1698 
 - ~~[ ] This currently works by setting the state of the view unminimized for the duration of scale; this is the easiest option to make it show up without extra hacks. Still, it could be an issue whether it is "logically" minimized or not (e.g. if a taskbar is displaying minimized views separately). Also, setting the minimized state automatically adjusts focus ([here](https://github.com/WayfireWM/wayfire/blob/84ec9fce347c330ad04c30036d22d3367e915802/src/view/view.cpp#L397)), which is not really needed here.~~

ammen99 commented:
This needs a rebase, otherwise is a good idea.

I think it makes the most sense to have this as an option in scale itself.
As to the minimized state: now there is a separation between logically minimized (view->minimized), and to hide/show the view its root node is enabled or disabled, so you could just enable and then disable the root node to implement this.

dkondor commented:
Hi,

I gave this a try, since it seemed simple, but I cannot get it working. My attempt is in the [scale_minimized3](https://github.com/dkondor/wayfire/tree/scale_minimized3) branch.
The problem is that when trying to enable or disable showing minimized views (lines 210 and 250), if I use `get_transformed_node()` (as the code for hiding the filtered views does), then the views still do not show up (there is an empty slot in their place, but cannot be selected with the mouse).
However, if I use `get_root_node()` (as it is now in my code), everything seems to work well at first, but if I exit scale by selecting a non-minimized view and then start scale again, the minimized views do not show up anymore. If I exit scale by selecting a minimized view (which becomes unminimized as a consequence), I cannot minimize it anymore.

I wonder if I'm missing something how the scenegraph interacts with the minimized state? I tried to look into the implementation (in toplevel-view.cpp), but I don't see anything obvious, apart from the minimized / disappeared signals that get emitted.

soreau commented:
> Hi,
> 
> I gave this a try, since it seemed simple, but I cannot get it working. My attempt is in the [scale_minimized3](https://github.com/dkondor/wayfire/tree/scale_minimized3) branch. The problem is that when trying to enable or disable showing minimized views (lines 210 and 250), if I use `get_transformed_node()` (as the code for hiding the filtered views does), then the views still do not show up (there is an empty slot in their place, but cannot be selected with the mouse). However, if I use `get_root_node()` (as it is now in my code), everything seems to work well at first, but if I exit scale by selecting a non-minimized view and then start scale again, the minimized views do not show up anymore. If I exit scale by selecting a minimized view (which becomes unminimized as a consequence), I cannot minimize it anymore.
> 
> I wonder if I'm missing something how the scenegraph interacts with the minimized state? I tried to look into the implementation (in toplevel-view.cpp), but I don't see anything obvious, apart from the minimized / disappeared signals that get emitted.

Hi, try this:
```
wf::scene::set_node_enabled(view->get_root_node(), true);
wf::scene::set_node_enabled(view->get_transformed_node(), true);
```

when adding the transformer. Hope this helps.

dkondor commented:
Hi,

thank you, this fixed my first problem (view not appearing the 2nd time scale is used). Also, I did not realize before that `set_node_enabled()` uses a counter internally. This made me come up with a solution for my second problem (view becomes unminimizable), but it seems very hackish.

Tracking down why sometimes I don't disable the root node when needed, I found that the code for minimizing a view [here](https://github.com/WayfireWM/wayfire/blob/f848c7569e785b00edc5464c40a316577741c488/src/view/toplevel-view.cpp#L214) always sends a `view_disappeared_signal`, regardless whether a view is being minimized or restored. However, in scale ([here](https://github.com/WayfireWM/wayfire/blob/f848c7569e785b00edc5464c40a316577741c488/plugins/scale/scale.cpp#L1081)), this always results in the view in question being removed from `scale_data`. However, in this case, I would really need to distinguish between a view appearing or disappearing -- the [solution](https://github.com/dkondor/wayfire/blob/57f21ba89eaf867018e9f604f064cb1d54823ddc/plugins/scale/scale.cpp#L1111) I could come up with so far is a bad hack that will lead to hard-to-reproduce crashes. Is there a better way to distinguish between views appearing and disappearing?

ammen99 commented:
> Tracking down why sometimes I don't disable the root node when needed, I found that the code for minimizing a view [here](https://github.com/WayfireWM/wayfire/blob/f848c7569e785b00edc5464c40a316577741c488/src/view/toplevel-view.cpp#L214) always sends a `view_disappeared_signal`, regardless whether a view is being minimized or restored.

Oh well, this is definitely an oversight, it should be emitted only when minimizing.

> Is there a better way to distinguish between views appearing and disappearing?

The disappear signal should never be called (current behavior is a bug) when a view 'appears', so I hope this answers your question?

dkondor commented:
Yes, it answers my question :) changing it solves the problem I'm having here as well. I'll open a separate PR for that. Thanks!

ammen99 reviewed (commented):
> Should minimized view be differentiated in some way? E.g. more faded, different title overlay, etc.

An alpha < 1.0 would probably make sense.

> Any view that is focused during scale is unminimzed.

Yeah, this is unfortunate. I think the best solution is probably to refactor scale to simply not change the focus at all, but keep track of the currently selected view manually, and focus the selected view only at the end.

dkondor commented on plugins/scale/scale.cpp:
```diff
@@ -198,6 +200,16 @@ class wayfire_scale : public wf::per_output_plugin_instance_t,
+        {
+            tr->alpha = 0.0;
+            wf::scene::set_node_enabled(view->get_root_node(), true);
+            wf::scene::set_node_enabled(view->get_transformed_node(), true);
```
Actually, no need for it -- I've added it before when experimenting

dkondor commented:
OK, I've addressed some of the remaining issues.

> > Should minimized view be differentiated in some way? E.g. more faded, different title overlay, etc.
> 
> An alpha < 1.0 would probably make sense.

I've added a configuration option for this; however, the difference with a "normal" unfocused view seems to be not that large to me subjectively, but this could be adjusted later.

> 
> > Any view that is focused during scale is unminimzed.
> 
> Yeah, this is unfortunate. I think the best solution is probably to refactor scale to simply not change the focus at all, but keep track of the currently selected view manually, and focus the selected view only at the end.

This is probably best addressed separately; I might have some time to work on this later this week.

ammen99 commented:
> I've added a configuration option for this; however, the difference with a "normal" unfocused view seems to be not that large to me subjectively, but this could be adjusted later.

Personally, I would make such views much fainter (say alpha = 0.3), so an option sounds good :)

ammen99 commented on plugins/scale/scale.cpp:
```diff
@@ -860,7 +885,9 @@ class wayfire_scale : public wf::per_output_plugin_instance_t,
             {
-                return wf::get_focus_timestamp(a) > wf::get_focus_timestamp(b);
+                /* note: avoid focusing minimized views if possible */
+                return (a->minimized < b->minimized) ||
```
I would split this into an if - else for better readability.

dkondor replied:
I've come up with an alternative, since it's also a bit weird to compare bools this way.

ammen99 commented on plugins/scale/scale.cpp:
```diff
@@ -1082,15 +1112,27 @@ class wayfire_scale : public wf::per_output_plugin_instance_t,
         {
-            remove_view(view);
-            if (scale_data.empty())
+            if (include_minimized && view->minimized && should_scale_view(view))
```
I think we should be really careful here, view_disappeared is not emitted only when the view is minimized. This path here could be triggered if for example a minimized view is closed (for example by killing the client).

I tested this locally and it doesn't actually cause any issues because `should_scale_view()` returns false for unmapped views - but I thought it might be worth adding a comment for future reference :)

dkondor replied:
Yes, I've run into this issue while testing -- I was experimenting if there is a more efficient way to distinguish between these two cases than `should_scale_view()`, but decided that this is likely the most robust one.

ammen99 reviewed (approved):
Two small things, otherwise I tested in some complex scenarios and it always works, so good job and thanks :)
